test/metric_learn_test.py

```diff
@@ -76,6 +77,124 @@ def test_singular_returns_pseudo_inverse(self):
                     pseudo_inverse)
 
 
+class TestSCML(MetricTestCase):
+  def test_iris(self):
+    scml = SCML_Supervised()
+    scml.fit(self.iris_points, self.iris_labels)
+
+    csep = class_separation(scml.transform(self.iris_points), self.iris_labels)
+    self.assertLess(csep, 0.3)
+
+  def test_bad_basis(self):
+    scml = SCML(basis='bad_basis')
+    triplets = np.ones((3, 3, 3))
+    authorized_basis = ['triplet_diffs']
+    msg = ("`basis` must be one of the options '{}' or an array of shape "
+           "(n_basis, n_features).".format("', '".join(authorized_basis)))
+    with pytest.raises(ValueError) as raised_error:
+      scml.fit(triplets)
+    assert msg == raised_error.value.args[0]
+
+  def test_big_n_basis(self):
+    scml = SCML(n_basis=4)
+    triplets = np.random.rand(3, 3, 3)
+    n_basis = 3
+    msg = ("The selected number of basis is greater than the number of points"
+           ", only n_basis = %d will be generated" % n_basis)
+    with pytest.warns(UserWarning) as raised_warning:
+      scml.fit(triplets)
+    assert msg == raised_warning[0].message.args[0]
+
+  def test_n_basis_wrong_type(self):
+    X = np.array([[0, 0], [0, 1], [2, 0], [2, 1]])
+    y = np.array([1, 0, 1, 0])
+
+    constraints = Constraints(y)
+    triplets = constraints.generate_knntriplets(X, k_genuine=1, k_impostor=1)
+    triplets = X[triplets]
+
+    n_basis = 4.0
+
+    scml = SCML(n_basis=n_basis)
+    msg = ("n_basis should be an integer, instead it is of type %s"
+           % type(n_basis))
+    with pytest.raises(ValueError) as raised_error:
+      scml.fit(triplets)
+    assert msg == raised_error.value.args[0]
+
+  def test_bad_basis_supervised(self):
+    scml = SCML_Supervised(basis='bad_basis')
+    X = np.array([[0, 0], [0, 1], [2, 0], [2, 1]])
+    y = np.array([1, 0, 1, 0])
+    authorized_basis = ['triplet_diffs', 'LDA']
+    msg = ("`basis` must be one of the options '{}' or an array of shape "
+           "(n_basis, n_features).".format("', '".join(authorized_basis)))
+    with pytest.raises(ValueError) as raised_error:
+      scml.fit(X, y)
+    assert msg == raised_error.value.args[0]
+
+  def test_small_n_basis_supervised(self):
+    X = np.array([[0, 0], [1, 1], [3, 3]])
+    y = np.array([1, 2, 3])
+
+    labels, class_count = np.unique(y, return_counts=True)
+    n_class = len(labels)
+
+    scml = SCML_Supervised(n_basis=n_class)
+    msg = ("The number of basis should be greater than the number of classes")
+    with pytest.raises(ValueError) as raised_error:
+      scml.fit(X, y)
+    assert msg == raised_error.value.args[0]
+
+  def test_big_n_basis_supervised(self):
```

Comment:
That actually would work ok but as LDA builds the scalings for the maximum separation of each class, having less than the number of classes would leave classes with "no representation" in the basis set. So the error is merely because it would be a really bad decision, maybe I could replace it for a warning.

Reply:
Ok, makes sense. You need to make the message more informative (not clear what "clusters" are)

Reply:
Not sure what you meant there by clusters. But I just changed the Error section for the following:
```
    if n_basis < n_class:
      warnings.warn("The number of basis is less than the number of classes,"
                    " this will lead to less basis than the amount yielded by"
                    " LDA")
```
Let me know what you think about this alternative.

Reply:
It is not easy for the user to understand the consequences I think ("less basis than the amount yielded by LDA" - so what?). Maybe "The number of basis is less than the number of classes, which may lead to poor discriminative performance."? Or something of the same fashion to be more concrete

Regarding clusters, I guess I got confused with the next test, which has this rather confusing message
```
msg = ("The selected number of basis needs a greater number of clusters"
           " than the number of available samples")
```

Reply:
How you phrase it is indeed way more clearer! Yes, I get that the other message is kind of confusing but I didn't know how to phrase it better without having to explain fully how the basis are generated. I will give it another try now!

Reply:
I have looked at the new attempt. I think it is still very difficult to understand for the average user who does not even know what cluster refers too. Maybe you should just switch to something like
"Not enough samples to generate {} LDA bases: `n_basis` should be at most {}"
This way the user knows that there is a problem and how to fix it